scrapy/http/request/__init__.py

```diff
@@ -43,6 +57,22 @@ def __init__(self, url, callback=None, method='GET', headers=None, body=None,
         self._cb_kwargs = dict(cb_kwargs) if cb_kwargs else None
         self.flags = [] if flags is None else list(flags)
 
+    @classmethod
+    def from_dict(cls: Type[RequestType], d: dict, spider: Optional[SpiderType] = None) -> RequestType:
```

Comment:
Hmm.. This reads (and I think mypy understands it this way) as it returns an instance of cls, but it can return arbitrary class - the one saved when serializing. So FormRequest.from_dict can return an object of arbitrary class, not just FormRequest. I wonder if that was the reason to have deserialization as a function, not as a classmethod.

Reply:
Interesting, I hadn't thought about that. What do you think about [this changeset](https://github.com/scrapy/scrapy/compare/9ef7b99ba30ede7fc7227f75965423802e0c7dbd...bf62b608ed95b924eb17b46d91fbf97bf5342379)?

Reply:
I’m not sure about forcing `FormRequest` to be the return value when using `FormRequest.from_dict` (or any other request subclass), but not forcing it for `Request.from_dict`.

I think we should either returns an instance of whatever class was serialized, raise a ValueError if the serialized class does not match the `from_dict` class (which would probably be useless in scenarios where you cannot know if that’s the case beforehand), or we go back to having a function, at least for the `from_dict` part.

Reply:
I'd prefer to keep all logic within the classes. Not forcing a particular class from `Request` makes sense to me because it's the highest class in the hierarchy. Then again, I do think it makes sense to return a specific instance if a user invokes a specific class. I believe this difference gives us good flexibility, i.e. you want your own class: call `YourRequestClass.from_dict`, you want whatever the original request was, call `Request.from_dict`. It shouldn't be too confusing for users, but I can document it properly to make it explicit.